### ISSUE TITLE
AQL optimization issue with LET variable assignment

### ISSUE THREAD

**Issue by ropeladder** [arangodb/arangodb#2400]
## my environment running ArangoDB
I'm using the latest ArangoDB of the respective release series:
- [X] 3.1

On this operating system:
- [X] Linux 
  - [ ] Ubuntu .deb

### this is an AQL-related issue:
[X] I'm using graph features

I'm issuing AQL via:
- [X] web interface with this browser: Firefox    running on this OS: Linux Mint 18.1

The AQL query in question is:
```
LET allMatches = (
FOR node, edge, path IN 1..3 ANY "[vertex_id]" GRAPH "[my_graph]"
  RETURN path
  )
LET shortestPathLength = (MIN(FOR m IN allMatches RETURN LENGTH(m.edges)))
//LET shortestPathLength = MIN(FOR m IN allMatches COLLECT len = LENGTH(m.edges) RETURN len)
FOR path IN allMatches
  FILTER LENGTH(path.edges) == shortestPathLength // 1
  RETURN path
```
The query above is extremely slow (53 seconds) using a dense graph, possibly because it appears to be re-running the `MIN()` function for every iteration of the main (second) `FOR` loop. If I run the same query using the second `LET shortestPathLength` statement instead of the first, or I substitute in `1` for `shortestPathLength` in the `FILTER` line, the query takes under one second.

I made a more easily reproducible example (obviously this is not the smartest way to find a minimum, my point it just that it doesn't make sense that this query is so slow):
```
LET allNums = (FOR x IN 1..30000 RETURN RAND())
LET minNum = MIN(allNums)
FOR num IN allNums
  FILTER num == minNum
  RETURN num

**Comment by jsteemann** [comment]
It looks like the optimizer has moved the statement `LET minNum = MIN(allNums)` from the latter query into the `FOR` loop. It did so to get rid of one variable. But by this query became
```
LET allNums = (FOR x IN 1..30000 RETURN RAND())
FOR num IN allNums
  FILTER num == MIN(allNums)
  RETURN num 
```
which means the `MIN()` function call was executed repeatedly, inside the `FOR` loop. This is worse than executing it once outside the loop.
The optimizer has now been adjusted to not move expressions containing function calls into loops when merging them with other existing expressions.

**Comment by jsteemann** [comment]
Fixed in 3.1.17 and devel.

**Comment by dothebart** [comment]
ArangoDB 3.1.17 containing this fix is available for download now.
